pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/SortedMailboxReceiveOperator.java

```diff
@@ -0,0 +1,189 @@
+ * {@link MultiStageOperator#getNextBlock()}()} API in a sorted manner.
+ *
+ *  TODO: Once sorting on the {@code MailboxSendOperator} is available, modify this to use a k-way merge instead of
+ *        resorting via the PriorityQueue.
+ */
+public class SortedMailboxReceiveOperator extends BaseMailboxReceiveOperator {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SortedMailboxReceiveOperator.class);
+  private static final String EXPLAIN_NAME = "SORTED_MAILBOX_RECEIVE";
+
+  private final List<RexExpression> _collationKeys;
+  private final List<RelFieldCollation.Direction> _collationDirections;
+  private final boolean _isSortOnSender;
+  private final boolean _isSortOnReceiver;
+  private final DataSchema _dataSchema;
+  private final PriorityQueue<Object[]> _priorityQueue;
+  private boolean _isSortedBlockConstructed;
+
+  public SortedMailboxReceiveOperator(OpChainExecutionContext context, RelDistribution.Type exchangeType,
+      List<RexExpression> collationKeys, List<RelFieldCollation.Direction> collationDirections, boolean isSortOnSender,
+      boolean isSortOnReceiver, DataSchema dataSchema, int senderStageId, int receiverStageId) {
+    this(context, context.getMetadataMap().get(senderStageId).getServerInstances(), exchangeType, collationKeys,
+        collationDirections, isSortOnSender, isSortOnReceiver, dataSchema, senderStageId,
+        receiverStageId, context.getTimeoutMs());
+  }
+
+  // TODO: Move deadlineInNanoSeconds to OperatorContext.
+  // TODO: Remove boxed timeoutMs value from here and use long deadlineMs from context.
+  public SortedMailboxReceiveOperator(OpChainExecutionContext context, List<VirtualServer> sendingStageInstances,
+      RelDistribution.Type exchangeType, List<RexExpression> collationKeys,
+      List<RelFieldCollation.Direction> collationDirections, boolean isSortOnSender, boolean isSortOnReceiver,
+      DataSchema dataSchema, int senderStageId, int receiverStageId, Long timeoutMs) {
+    super(context, sendingStageInstances, exchangeType, senderStageId, receiverStageId, timeoutMs);
+    _collationKeys = collationKeys;
+    _collationDirections = collationDirections;
+    _isSortOnSender = isSortOnSender;
+    _isSortOnReceiver = isSortOnReceiver;
+    _dataSchema = dataSchema;
+    Preconditions.checkState(!CollectionUtils.isEmpty(collationKeys) && isSortOnReceiver,
+        "Collation keys should exist and sorting must be enabled otherwise use non-sorted MailboxReceiveOperator");
+    _priorityQueue = new PriorityQueue<>(new SortUtils.SortComparator(collationKeys, collationDirections,
+        dataSchema, false));
+    _isSortedBlockConstructed = false;
+  }
+
+  @Nullable
+  @Override
+  public String toExplainString() {
+    return EXPLAIN_NAME;
+  }
+
+  @Override
+  protected TransferableBlock getNextBlock() {
+    if (_upstreamErrorBlock != null) {
+      cleanUpResourcesOnError();
```

Comment:
Can we do all clean-up in the cancel/close calls?

Reply:
done